File: LibretroRT.FrontendComponents.Win2DRenderer/RenderTargetManager.cs
using Microsoft.Graphics.Canvas;
using System;
using System.Collections.Generic;
using Windows.Foundation;
using Windows.Graphics.DirectX;

namespace LibretroRT.FrontendComponents.Win2DRenderer
{
    internal class RenderTargetManager : IDisposable
    {
        private const uint RenderTargetMinSize = 1024;
        private const DirectXPixelFormat RenderTargetPixelFormat = DirectXPixelFormat.B8G8R8A8UIntNormalized;
        private const int RenderTargetPixelSize = 4;

        private static readonly IReadOnlyDictionary<PixelFormats, int> PixelFormatsSizeMapping = new Dictionary<PixelFormats, int>
        {
            { PixelFormats.FormatXRGB8888, 4 },
            { PixelFormats.FormatRGB565, 2 },
            { PixelFormats.Format0RGB1555, 2 },
            { PixelFormats.FormatUknown, 0 },
        };

        private readonly object RenderTargetLock = new object();
        private CanvasBitmap RenderTarget = null;
        private byte[] RenderTargetBuffer = null;
        private Rect RenderTargetViewport = new Rect();
        //This may be different from viewport's width/haight.
        private float RenderTargetAspectRatio = 1.0f;

        private PixelFormats currentCorePixelFormat = PixelFormats.FormatUknown;
        public PixelFormats CurrentCorePixelFormat
        {
            get { return currentCorePixelFormat; }
            set { currentCorePixelFormat = value; CurrentCorePixelSize = PixelFormatsSizeMapping[currentCorePixelFormat]; }
        }

        private int CurrentCorePixelSize = 0;

        public void Dispose()
        {
            RenderTarget?.Dispose();
            RenderTarget = null;
        }

        public void Render(CanvasDrawingSession drawingSession, Size canvasSize)
        {
            var viewportWidth = RenderTargetViewport.Width;
            var viewportHeight = RenderTargetViewport.Height;

            if (RenderTarget == null || viewportWidth <= 0 || viewportHeight <= 0)
                return;

            lock (RenderTargetLock)
            {
                var destinationRect = ComputeBestFittingSize(canvasSize, RenderTargetAspectRatio);
                drawingSession.DrawImage(RenderTarget, destinationRect, RenderTargetViewport);
            }
        }

        public void UpdateFromCoreOutput(byte[] frameBuffer, uint width, uint height, uint pitch)
        {
            if (frameBuffer == null || RenderTarget == null || CurrentCorePixelSize == 0)
                return;

            lock (RenderTargetLock)
            {
                var virtualWidth = pitch / CurrentCorePixelSize;
                RenderTargetViewport.Width = width;
                RenderTargetViewport.Height = height;

                switch(CurrentCorePixelFormat)
                {
                    case PixelFormats.FormatXRGB8888:
                        RenderTarget.SetPixelBytes(frameBuffer, 0, 0, (int)virtualWidth, (int)height);
                        break;
                    case PixelFormats.FormatRGB565:
                        ColorConverter.ConvertFrameBufferRGB565ToXRGB8888(frameBuffer, width, height, pitch, RenderTargetBuffer);
                        RenderTarget.SetPixelBytes(RenderTargetBuffer, 0, 0, (int)width, (int)height);
                        break;
                }
            }
        }

        public void UpdateRenderTargetSize(ICanvasResourceCreator resourceCreator, GameGeometry geometry)
        {
            RenderTargetAspectRatio = geometry.AspectRatio;
            if (RenderTargetAspectRatio < 0.1f)
            {
                RenderTargetAspectRatio = (float)(geometry.BaseWidth) / geometry.BaseHeight;
            }

            if (RenderTarget != null)
            {
                var currentSize = RenderTarget.Size;
                if (currentSize.Width >= geometry.MaxWidth && currentSize.Height >= geometry.MaxHeight)
                {
                    return;
                }
            }

            lock (RenderTargetLock)
            {
                var size = Math.Max(Math.Max(geometry.MaxWidth, geometry.MaxHeight), RenderTargetMinSize);
                size = ClosestGreaterPowerTwo(size);
                RenderTargetBuffer = new byte[size * size * RenderTargetPixelSize];

                RenderTarget?.Dispose();
                RenderTarget = CanvasBitmap.CreateFromBytes(resourceCreator, RenderTargetBuffer, (int)size, (int)size, RenderTargetPixelFormat);
            }
        }

        private static Rect ComputeBestFittingSize(Size viewportSize, float aspectRatio)
        {
            Rect output;
            var candidateWidth = Math.Floor(viewportSize.Height * aspectRatio);
            if (viewportSize.Width >= candidateWidth)
            {
                var size = new Size(candidateWidth, viewportSize.Height);
                output = new Rect(new Point((viewportSize.Width - candidateWidth) / 2, 0), size);
            }
            else
            {
                var height = viewportSize.Width / aspectRatio;
                var size = new Size(viewportSize.Width, height);
                output = new Rect(new Point(0, (viewportSize.Height - height) / 2), size);
            }

            return output;
        }

        private static uint ClosestGreaterPowerTwo(uint value)
        {
            uint output = 1;
            while (output < value)
            {
                output *= 2;
            }

            return output;
        }
    }
}

File: LibretroRT.FrontendComponents.Renderer/RenderTargetManager.cpp
#include "pch.h"
#include "RenderTargetManager.h"
#include "ColorConverter.h"

using namespace LibretroRT_FrontendComponents_Renderer;

using namespace Windows::Graphics::DirectX::Direct3D11;

RenderTargetManager::RenderTargetManager(CanvasAnimatedControl^ canvas) :
	Canvas(canvas),
	OpenGLESManager(OpenGLES::GetInstance())
{
	__abi_ThrowIfFailed(GetDXGIInterface(canvas->Device, Direct3DDevice.GetAddressOf()));

	ColorConverter::InitializeLookupTable();
}


RenderTargetManager::~RenderTargetManager()
{
	DestroyRenderTargets();
}

void RenderTargetManager::UpdateFormat()
{
	if (Geometry == nullptr)
	{
		return;
	}

	critical_section::scoped_lock lock(RenderTargetCriticalSection);

	bool shouldUpdate = true;
	if (Win2DTexture)
	{
		auto size = Win2DTexture->SizeInPixels;
		shouldUpdate = (size.Width < geometry->MaxWidth || size.Height < geometry->MaxHeight);
	}

	if (shouldUpdate)
	{
		auto dimension = max(geometry->MaxWidth, geometry->MaxHeight);
		dimension = max(dimension, RenderTargetMinSize);
		dimension = ClosestGreaterPowerTwo(dimension);
		CreateRenderTargets(Canvas, dimension, dimension);
	}
}

void RenderTargetManager::UpdateFromCoreOutput(const Array<byte>^ frameBuffer, unsigned int width, unsigned int height, unsigned int pitch)
{
	RenderTargetViewport.Width = width;
	RenderTargetViewport.Height = height;

	critical_section::scoped_lock lock(RenderTargetCriticalSection);
	
	if (usingHardwareRendering)
	{
		OpenGLESManager->MakeCurrent(OpenGLESSurface);
		glClearColor(1.0f, 0.0f, 0.0f, 1.0f);
		glClear(GL_COLOR_BUFFER_BIT);
		glFlush();
	}
	else
	{
		ComPtr<ID3D11DeviceContext> d3dContext;
		Direct3DDevice->GetImmediateContext(d3dContext.GetAddressOf());

		D3D11_MAPPED_SUBRESOURCE mappedResource;
		__abi_ThrowIfFailed(d3dContext->Map(Direct3DTexture.Get(), 0, D3D11_MAP_WRITE_DISCARD, 0, &mappedResource));

		auto pSrc = frameBuffer->Data;
		auto pDest = (byte*)mappedResource.pData;

		switch (PixelFormat)
		{
		case PixelFormats::FormatXRGB8888:
		{
			for (auto i = 0; i < height; i++)
			{
				static const size_t rgba8888bpp = 4;
				memcpy(pDest, pSrc, width * rgba8888bpp);

				pSrc += pitch;
				pDest += mappedResource.RowPitch;
			}
		}
		break;
		case PixelFormats::FormatRGB565:
		{
			for (auto i = 0; i < height; i++)
			{
				ColorConverter::ConvertRGB565ToXRGB8888(pDest, pSrc, width);

				pSrc += pitch;
				pDest += mappedResource.RowPitch;
			}			
		}
		break;
		}
		
		d3dContext->Unmap(Direct3DTexture.Get(), 0);
	}
}

void RenderTargetManager::Render(CanvasDrawingSession^ drawingSession, Size canvasSize)
{
	if (Win2DTexture == nullptr || RenderTargetViewport.Width <= 0 || RenderTargetViewport.Height <= 0)
	{
		return;
	}

	critical_section::scoped_lock lock(RenderTargetCriticalSection);

	auto destinationRect = ComputeBestFittingSize(canvasSize, Geometry->AspectRatio);
	drawingSession->DrawImage(Win2DTexture, destinationRect, RenderTargetViewport);
}

void RenderTargetManager::CreateRenderTargets(CanvasAnimatedControl^ canvas, unsigned int width, unsigned int height)
{
	DestroyRenderTargets();

	ComPtr<IDXGISurface> d3dSurface;

	if (usingHardwareRendering)
	{
		OpenGLESSurface = OpenGLESManager->CreateSurface(width, height, EGL_TEXTURE_RGBA);
		auto surfaceHandle = OpenGLESManager->GetSurfaceShareHandle(OpenGLESSurface);

		__abi_ThrowIfFailed(Direct3DDevice->OpenSharedResource(surfaceHandle, __uuidof(IDXGISurface), (void**)d3dSurface.GetAddressOf()));
	}
	else
	{
		D3D11_TEXTURE2D_DESC texDesc = { 0 };
		texDesc.Width = width;
		texDesc.Height = height;
		texDesc.Format = DXGI_FORMAT_B8G8R8A8_UNORM;
		texDesc.MipLevels = 1;
		texDesc.ArraySize = 1;
		texDesc.SampleDesc.Count = 1;
		texDesc.SampleDesc.Quality = 0;
		texDesc.Usage = D3D11_USAGE_DYNAMIC;
		texDesc.BindFlags = D3D11_BIND_SHADER_RESOURCE;
		texDesc.CPUAccessFlags = D3D11_CPU_ACCESS_WRITE;
		texDesc.MiscFlags = 0;

		__abi_ThrowIfFailed(Direct3DDevice->CreateTexture2D(&texDesc, nullptr, Direct3DTexture.GetAddressOf()));
		__abi_ThrowIfFailed(Direct3DTexture.As(&d3dSurface));
	}

	auto winRTSurface = CreateDirect3DSurface(d3dSurface.Get());
	Win2DTexture = CanvasBitmap::CreateFromDirect3D11Surface(canvas->Device, winRTSurface);
}

void RenderTargetManager::DestroyRenderTargets()
{
	Win2DTexture = nullptr;
	Direct3DTexture.Reset();

	if (OpenGLESSurface != EGL_NO_SURFACE)
	{
		OpenGLESManager->DestroySurface(OpenGLESSurface);
		OpenGLESSurface = EGL_NO_SURFACE;
	}
}

Rect RenderTargetManager::ComputeBestFittingSize(Size viewportSize, float aspectRatio)
{
	auto candidateWidth = std::floor(viewportSize.Height * aspectRatio);
	if (viewportSize.Width >= candidateWidth)
	{
		Size size(candidateWidth, viewportSize.Height);
		Rect output(Point((viewportSize.Width - candidateWidth) / 2, 0), size);
		return output;
	}
	else
	{
		auto height = viewportSize.Width / aspectRatio;
		Size size(viewportSize.Width, height);
		Rect output(Point(0, (viewportSize.Height - height) / 2), size);
		return output;
	}
}

unsigned int RenderTargetManager::ClosestGreaterPowerTwo(unsigned int value)
{
	unsigned int output = 1;
	while (output < value)
	{
		output *= 2;
	}

	return output;
}